org.scala-ide.sdt.core/src/org/scalaide/core/internal/project/scopes/BuildScopeUnit.scala

```diff
@@ -0,0 +1,90 @@
+package org.scalaide.core.internal.project.scopes
+
+import org.eclipse.core.resources.IFile
+import org.eclipse.core.resources.IResource
+import org.eclipse.core.runtime.SubMonitor
+import org.scalaide.core.IScalaPlugin
+import org.scalaide.core.IScalaProject
+import org.scalaide.core.SdtConstants
+import org.scalaide.core.internal.builder.BuildProblemMarker
+import org.scalaide.core.internal.project.UnitBuilder
+import org.scalaide.ui.internal.preferences.ScalaPluginSettings
+import org.scalaide.util.internal.SettingConverterUtil
+import org.scalaide.core.internal.project.CompileScope
+import org.eclipse.core.resources.IMarker
+import org.eclipse.jdt.core.compiler.IProblem
+import org.eclipse.jdt.ui.text.java.IProblemLocation
+import org.eclipse.jdt.core.IJavaProject
+import org.eclipse.jdt.core.IJavaModelMarker
+
+case object CompileMacrosScope extends CompileScope {
+  def name = "Macros"
+}
+case object CompileMainScope extends CompileScope {
+  def name = "Main"
+}
+case object CompileTestsScope extends CompileScope {
+  def name = "Tests"
+}
+
+abstract class BuildScopeUnit(val owningProject: IScalaProject, val unitBuilder: UnitBuilder,
+    val dependentUnitInstances: Seq[BuildScopeUnit] = Seq.empty) {
+  @volatile private var error: Boolean = false
+
+  def isValidSourcePath(path: String): Boolean
+
+  def dependentUnits: Seq[CompileScope]
+
+  def scope: CompileScope
+
+  def hasError: Boolean = error
+
+  def compileUnit(addedOrUpdated: Set[IFile], removed: Set[IFile], monitor: SubMonitor): Unit = {
+    val getJavaSourcesOfThisScope: Unit => Set[IFile] = (Unit) => {
+      val Dot = 1
+      getToCompileFor(owningProject.allSourceFiles
+        .filter { _.getLocation.getFileExtension == SdtConstants.JavaFileExtn.drop(Dot) })
+    }
+    val resetJavaMarkers: Set[IFile] => Set[IFile] = javaFiles => {
```

Comment:
Why do we need to do this? Previously the builder removed all error markers for a given project, but only the Scala error markers. Now we need to touch Java markers, which is strange.